src/main/java/org/traccar/database/OpenIdProvider.java

```diff
@@ -69,28 +77,51 @@ public class OpenIdProvider {
     private final String adminGroup;
+    private final String allowGroup;
 
     private LoginService loginService;
 
     @Inject
-    public OpenIdProvider(Config config, LoginService loginService) {
-        this.loginService = loginService;
-
-        force = config.getBoolean(Keys.OPENID_FORCE);
-        clientId = new ClientID(config.getString(Keys.OPENID_CLIENTID));
-        clientAuth = new ClientSecretBasic(clientId, new Secret(config.getString(Keys.OPENID_CLIENTSECRET)));
-
-        try {
-            callbackUrl = new URI(config.getString(Keys.WEB_URL, "") + "/api/session/openid/callback");
-            authUrl = new URI(config.getString(Keys.OPENID_AUTHURL, ""));
-            tokenUrl = new URI(config.getString(Keys.OPENID_TOKENURL, ""));
-            userInfoUrl = new URI(config.getString(Keys.OPENID_USERINFOURL, ""));
-            baseUrl = new URI(config.getString(Keys.WEB_URL, ""));
+    public OpenIdProvider(
+        Config config, LoginService loginService, HttpClient httpClient, ObjectMapper objectMapper
+        ) throws InterruptedException, IOException {
+            this.loginService = loginService;
+
+            force = config.getBoolean(Keys.OPENID_FORCE);
+            clientId = new ClientID(config.getString(Keys.OPENID_CLIENT_ID));
+            clientAuth = new ClientSecretBasic(clientId, new Secret(config.getString(Keys.OPENID_CLIENT_SECRET)));
+
+            try {
+                callbackUrl = new URI(config.getString(Keys.WEB_URL, "") + "/api/session/openid/callback");
+                baseUrl = new URI(config.getString(Keys.WEB_URL, ""));
+
+                if (config.hasKey(Keys.OPENID_ISSUER_URL)) {
+                    HttpRequest httpRequest = HttpRequest.newBuilder(
+                        URI.create(config.getString(Keys.OPENID_ISSUER_URL) + "/.well-known/openid-configuration"))
+                        .header("Accept", "application/json")
+                        .build();
+
+                    String httpResponse = httpClient.send(httpRequest, BodyHandlers.ofString()).body();
+
+                    Map<String, Object> discoveryMap = objectMapper.readValue(
+                        httpResponse, new TypeReference<Map<String, Object>>() { });
+
+                    authUrl = new URI((String) discoveryMap.get("authorization_endpoint"));
+                    tokenUrl = new URI((String) discoveryMap.get("token_endpoint"));
+                    userInfoUrl = new URI((String) discoveryMap.get("userinfo_endpoint"));
+
+                    LOGGER.info("OpenID Connect auto discovery successful");
+                } else {
+                    authUrl = new URI(config.getString(Keys.OPENID_AUTH_URL));
+                    tokenUrl = new URI(config.getString(Keys.OPENID_TOKEN_URL));
+                    userInfoUrl = new URI(config.getString(Keys.OPENID_USERINFO_URL));
+                }
         } catch (URISyntaxException error) {
             LOGGER.error("Invalid URIs provided in OpenID configuration");
```

Comment:
I would probably keep this exception as well. An exception will actually be more useful because it will include line numbers.

Reply:
And we can probably remove try catch